abmarl/trainers/base.py

```diff
@@ -0,0 +1,218 @@
+            for agent_id in obs
+        }
+
+    # TODO: Upgrade to generate_batch
+    def generate_episode(self, horizon=200, **kwargs):
+        """
+        Generate an episode of data.
+
+        The fundamental data object is a SAR--a (state, action, reward) tuple.
+        We restart the sim, generating initial observations (states) for agents
+        reporting from the sim. Then we use the compute_action function to generate
+        actions for those reporting agents. Those actions are given to the sim,
+        which steps forward and generates rewards and new observations for reporting
+        agents. This loop continues until the simulation is done or we hit the
+        horizon.
+
+        Args:
+            horizon: The maximum number of steps per epsidoe. They episode may
+                finish early, but it will not progress further than this number
+                of steps.
+
+        Returns:
+            Three dictionaries, one for observations, another for actions, and
+            another for rewards, thus making up the SAR sequences. The data is
+            organized by agent_id, so you would call
+                {observations, actions, rewards}[agent_id][i]
+            in order to extract the ith SAR for an agent.
+            NOTE: In multiagent simulations, the number of SARs may differ for
+            each agent.
+        """
+        # Reset the simulation and policies
+        obs = self.sim.reset()
+        done = {agent: False for agent in obs}
+        for policy in self.policies.values():
+            policy.reset()
+
+        # Data collection
+        observations, actions, rewards = {}, {}, {}
+        for agent_id, agent_obs in obs.items():
+            observations[agent_id] = [agent_obs]
+
+        # Generate episode of data
+        for j in range(horizon):
+            action = self.compute_actions(obs)
+            obs, reward, done, _ = self.sim.step(action)
+
+            # Store the data
+            for agent_id, agent_obs in obs.items():
+                try:
+                    observations[agent_id].append(agent_obs)
+                except KeyError:
+                    observations[agent_id] = [agent_obs]
+            for agent_id, agent_reward in reward.items():
+                try:
+                    rewards[agent_id].append(agent_reward)
+                except KeyError:
+                    rewards[agent_id] = [agent_reward]
+            for agent_id, agent_action in action.items():
+                try:
+                    actions[agent_id].append(agent_action)
+                except KeyError:
+                    actions[agent_id] = [agent_action]
+
+            # Exit if we're done
+            if done['__all__']:
+                break
+
+            # We should not request actions for any more done agents.
+            for agent_id, agent_done in done.items():
+                if agent_done:
+                    del obs[agent_id]
```

Comment:
Add a note: this is okay to delete because it has already been stored in observations